server/src/main/java/org/opensearch/common/path/PathTrie.java

```diff
@@ -405,4 +406,45 @@ public T next() {
             }
         };
     }
+
+    public Iterator<T> retrieveAll() {
+        Stack<TrieNode> stack = new Stack<>();
+        stack.add(root);
+
+        return new Iterator<T>() {
+            @Override
+            public boolean hasNext() {
+                while (!stack.empty()) {
+                    TrieNode node = stack.peek();
+
+                    if (node.value != null) {
+                        return true;
+                    }
+
+                    advance();
```

Comment:
Do you need `advance()` call here? It will change the node, right?

Reply:
I have to advance skipping `null` returns, so it means doing it in both `hasNext()` and `next()`. If you know of a more "elegant" way, I'm all ears :)

Reply:
Yeah, I realized that you have to skip `null` on the property, not on element, another solution I was thinking is pre-filter - never populate the stack with nodes that have `node.value == null` (but I haven't prototyped it)

Reply:
Well, I think you still will need to keep a stack for that approach.

The regular approach to this problem is to traverse recursively and iterate over all child nodes at every level. It's easy to do that if you gather nodes of interest while you do that, but that defeats the purpose of an iterator. A proper DFS/BFS iterator would traverse at the same time as it yields results, and resumes where it left off every time you call `next()`. I only know how to do that in Ruby :)

Reply:
> Well, I think you still will need to keep a stack for that approach.

Yes, that won't change I think but when you add to the stack (`advance`), the `node.children.values()` could be pre-filtered. So from the iterator perspective - nothing should change, it is still same lazy approach (at least this is what I had in mind). But I am perfectly fine with the way you do it now.

Reply:
You can't do that because you need to recurse deep into those children. You can have lots of nesting with `null` values.